pycc/rt/rtcc.py

```diff
@@ -292,6 +275,67 @@ def lagrangian(self, t, t1, t2, l1, l2):
         etwo = oooo_energy + vvvv_energy + ooov_energy + vvvo_energy + ovov_energy + oovv_energy
         return eref + eone + etwo
 
+    def phase(self, F, t1, t2):
+        """
+        Parameters
+        ----------
+        F : NumPy array
+            current (field-dependent Fock operator
+        t1, t2: NumPy arrays
+            current cluster amplitudes
+
+        Returns
+        -------
+        phase: scalar
+            wave function quasienergy/phase-factor with contribution defined as = exp(-phase(t))
+        """
+        contract = self.contract
+        o = self.ccwfn.o
+        v = self.ccwfn.v
+        L = self.ccwfn.H.L
+
+        if isinstance(t1, torch.Tensor):
+            eref = 2.0 * torch.trace(F[o,o])
+            tmp = self.ccwfn.H.L[o,o,o,o].to(self.ccwfn.device1)
+            eref -= torch.trace(opt_einsum.contract('i...i', tmp.swapaxes(0,1), backend='torch'))
+            del tmp
+
+        else:
+            eref = 2.0 * np.trace(F[o,o])
+            eref -= np.trace(np.trace(self.ccwfn.H.L[o,o,o,o], axis1=1, axis2=3))
+
+        if self.ccwfn.model == 'CCD':
+            ecc = contract('ijab,ijab->', t2, L[o,o,v,v])
+        else:
+            ecc = 2.0 * contract('ia,ia->', F[o,v], t1)
+            ecc += contract('ijab,ijab->', self.ccwfn.build_tau(t1, t2), L[o,o,v,v])
+
+        return (eref + ecc) * (-1.0j)
+
+    def autocorrelation(self, y_left, y_right):
+        contract = opt_einsum.contract
+
+        t1_l, t2_l, l1_l, l2_l, phase_l = self.extract_amps(y_left)
+        t1_r, t2_r, l1_r, l2_r, phase_r = self.extract_amps(y_right)
+
+        A = 1
+        A += contract("ia,ia->", l1_l, (t1_r - t1_l))
+        A += 0.5*contract("ijab,ijab->", l2_l, (t2_r - t2_l))
+        A += 0.5*contract("ijab,ia,jb->", l2_l, t1_l, t1_l)
+        A += 0.5*contract("ijab,ia,jb->", l2_l, t1_r, t1_r)
+        A -= contract("ijab,ia,jb->", l2_l, t1_l, t1_r)
+        A *= np.exp(-phase_l) * np.exp(phase_r)
+
+        B = 1
+        B -= contract("ia,ia->", l1_r, (t1_r - t1_l))
+        B -= 0.5*contract("ijab,ijab->", l2_r, (t2_r - t2_l))
+        B += 0.5*contract("ijab,ia,jb->", l2_r, t1_r, t1_r)
+        B += 0.5*contract("ijab,ia,jb->", l2_r, t1_l, t1_l)
+        B -= contract("ijab,ia,jb->", l2_r, t1_l, t1_r)
+        B *= np.exp(-phase_r) * np.exp(phase_l)
+
+        return 0.5*A + 0.5*np.conj(B)
```

Comment:
is this computing a symmetrized version of Eq. 19 from [J. Chem. Phys. 150, 144106 (2019)](https://doi.org/10.1063/1.5085390)?

Reply:
No, it's actually Eq. (18).  Eq. (19) is just used in the paper to make an argument about the connection between the autocorrelation function and the ground-state population.